torch/csrc/jit/codegen/cuda/scheduler_registry.h

```diff
@@ -0,0 +1,81 @@
+#pragma once
+
+#include <torch/csrc/jit/codegen/cuda/fusion.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+//! Virtual base class for schedule heuristics
+//!   heuristic implementations derive from this
+//!   class and implement a schedule(Fusion*)
+//!   and a bool canSchedule(Fusion*) interface
+class TORCH_CUDA_API SchedulerEntry {
+ public:
+  //! Fusion runtime facing API,
+  //!   builds a new entry with the given heuristics
+  //!   corresponding to the given fusion
+  static std::unique_ptr<SchedulerEntry> makeEntry(
+      ScheduleHeuristic sh,
+      Fusion* fusion,
+      ExpressionEvaluator& ee);
+
+  //! Fusion segmenter facing API,
+  //!   returns a schedule that applies in the given fusion, returns a nullopt
+  //!   if no schedule in the registry can handle.
+  static c10::optional<ScheduleHeuristic> proposeHeuristics(Fusion* fusion);
+
+  //! Fusion runtime facing API,
+  //!   schedule the given fusion with heuristics owned
+  //!   by this entry, for actual heuristics to override
+  virtual void schedule(Fusion* fusion) = 0;
+
+  //! Heuristic comparison
+  bool operator==(const SchedulerEntry& other) {
+    return sameAs(&other);
+  }
+
+  //! Heuristic comparison
+  bool sameAs(const SchedulerEntry* other);
```

Comment:
why do we need both op== and sameAs() - they seem to do the same thing?

Reply:
I actually was only using sameAs(). Thought op== would also be a useful short-cut. But removed it for now. Thanks for pointing this out 👍